jan2025/helpers.js

```diff
@@ -0,0 +1,369 @@
+      logs += `${timestamp} Updated project type value (node ${nodeId}); `;
+    }
+  });
+
+  return {
+    flowData: newFlowData,
+    logs: logs
+  };
+}
+
+const migrateSessionData = (sessionData) => {
+  const newsessionData = sessionData;
+
+  const passportData = sessionData?.passport?.data;
+  const breadcrumbs = sessionData?.breadcrumbs;
+
+  // Property types
+  if (Object.keys(passportData).includes("property.type")) {
+    newsessionData["passport"]["data"]["property.type"] = passportData["property.type"].map(
+      (current) => Object.keys(propertyTypeChanges).includes(current) ? propertyTypeChanges[current] : current
+    );
+
+    Object.entries(breadcrumbs).forEach(([nodeId, crumb]) => {
+      if (crumb.data?.hasOwnProperty("property.type")) {
+        newsessionData["breadcrumbs"][nodeId]["data"]["property.type"] = crumb["data"]["property.type"].map(
+          (current) => Object.keys(propertyTypeChanges).includes(current) ? propertyTypeChanges[current] : current
+        );
+      }
+    });
+  }
+
+  // Planning constraints (including `_constraints` audit key)
+  const combinedConstraintChanges = {...listedGradeChanges, ...floodZoneChanges};
+  if (Object.keys(passportData).includes("property.constraints.planning") || Object.keys(passportData).includes("_nots")) {
+    newsessionData["passport"]["data"]["property.constraints.planning"] = passportData?.["property.constraints.planning"]?.map(
+      (current) => Object.keys(combinedConstraintChanges).includes(current) ? combinedConstraintChanges[current] : current.replaceAll("article4", "articleFour")
+    );
+    newsessionData["passport"]["data"]["_nots"]["property.constraints.planning"] = passportData?.["_nots"]?.["property.constraints.planning"]?.map(
+      (current) => Object.keys(combinedConstraintChanges).includes(current) ? combinedConstraintChanges[current] : current.replaceAll("article4", "articleFour")
+    );
+
+    Object.entries(breadcrumbs).forEach(([nodeId, crumb]) => {
+      if (crumb.data?.hasOwnProperty("property.constraints.planning")) {
+        newsessionData["breadcrumbs"][nodeId]["data"]["property.constraints.planning"] = crumb["data"]["property.constraints.planning"].map(
+          (current) => Object.keys(combinedConstraintChanges).includes(current) ? combinedConstraintChanges[current] : current.replaceAll("article4", "articleFour")
+        );
+      }
+      if (crumb.data?.hasOwnProperty("_nots")) {
+        newsessionData["breadcrumbs"][nodeId]["data"]["_nots"]["property.constraints.planning"] = crumb["data"]["_nots"]["property.constraints.planning"].map(
+          (current) => Object.keys(combinedConstraintChanges).includes(current) ? combinedConstraintChanges[current] : current.replaceAll("article4", "articleFour")
+        );
+      }
+    });
+  }
+
+  if (Object.keys(passportData).includes("_constraints")) {
+    // _constraints is a list of two requests: planning data first, os roads second
```

Comment:
Super helpful comment - explains the `[0]` index.

Something to think about for future - maybe we should use on object (depend on fixed keys) here as opposed to an array. I could see us making well intentioned changes here and missing a subtlety like this that relies on the array order. Won't affect this PR in any way as it's perfect correct at this point in time.

https://github.com/theopensystemslab/planx-new/blob/main/editor.planx.uk/src/@planx/components/PlanningConstraints/Public.tsx#L140-L151